---
title: "BIOE591: Genomics for Ecology & Conservation"
format: html
---

**Course:** BIOE 591 (3 Credits)  
**Prerequisite:** BIOB480/BIOE548 or consent of instructor  
**Meeting Time:** T/TR 1:40 PM – 2:55 PM  
**Place:** 	Wilson 1-154  
**Instructor:** Dr. Ethan Linck (ethan.linck\@montana.edu; 406-994-2024)  
**Office Hours:** 1:00 PM – 4:00 Weds  
**Materials:** A laptop (tablets will not suffice). Windows users will need to download and install [Git for Windows](https://gitforwindows.org/) as a Unix shell emulator.   

## Catalog Description

Application of genomics methods to conservation, ecology, and evolution. Students gain familiarity with the basic tools and techniques of computational biology and genomics through case studies from the primary literature and analysis of empirical datasets. Examples and questions relevant to conservation biology and fish and wildlife management are emphasized.

## Overview

The development of affordable high-throughput sequencing technology in the 2000s and 2010s revolutionized the ability of molecular ecologists, wildlife and conservation biologists, and other environmental scientists to assay genome-wide DNA sequence variation from nonmodel organisms. Genomics methods are now commonly used to study everything from the dietary preferences of sturgeon to genes underpinning local adaptation in wolves, but can be difficult to apply or interpret without specialist training. This course provides a hands-on introduction to the use of genomics in ecology, conservation, and related fields. Students will gain hands-on experience with computational biology, analyzing empirical genomic datasets with commonly applied software on a high-performance computing cluster. Reading assignments from the primary literature will demonstrate the use of focal software packages while emphasizing case studies relevant to conservation biology and fish and wildlife management.  

## Course Organization and Format

The course will be offered as a mix of short lectures or demonstrations and discussions on reading assignments (typically Tuesdays) and supervised computer lab activities (typically Thursdays). 

## Learning Outcomes

-   Compare high-throughput sequencing approaches;
-   Evaluate genomics methods sections in the primary literature;
-   Apply basic tools and techniques from computational biology;
-   Analyze empirical datasets with genomics softwate;
-   Create scripts for custom data analysis tasks.

## Grading

Your grade will be calculated as the fraction of points earned out of a total of 200. 120 points will come from completing lab assignments (10 points each; turned in the following class period as homework), 60 will come from posting comments or questions in Canvas discussion threads on assigned reading (5 points each; engagement with 12/14 papers needed), and 20 points will come from participation (based on attendance, leading at least one paper discussion and participating in the classroom, and completing the assessment).

## Resources

I assume a solid grasp of basic genetics and evolution. If you consider yourself rusty in these areas, I recommend reviewing [the lecture notes from my Conservation Genetics course (BIOB480 / BIOE548)](https://elinck.org/popgen_congen_notes/). I take for granted that you will have computer skills typical of a first-year graduate student in the middle of the third decade of the 21st century. This means: proficiency with what Gotelli [calls the "unholy trinity"](https://gotellilab.github.io/Bio381/Lectures/Introduction.html) of M.S. Word, Excel and Powerpoint; a knowledge of basic computing hardware and operating systems; a knowledge of where to find things on your computer; and some background with scientific programming, even if shallow. 

Throughout the semester, you will likely find yourself resorting to Google, ChatGPT, and other resources to help run software and troubleshoot error messages. The following cheat sheets, tutorials and courses may also be of interest: 

- [Official GitHub Cheatsheet](https://education.github.com/git-cheat-sheet-education.pdf)  
- [Git Beginner Cheatsheet](https://mukulrathi.com/git-beginner-cheatsheet/) 
- [Command Line Cheatsheet #1](chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://adibro.github.io/Data-Science-Resources/Cheat-Sheets/CL-Git/Command-Line_Cheat-Sheet.pdf)
- [Command Line Cheatsheet #2](https://www.git-tower.com/blog/command-line-cheat-sheet)
- [Software Carpentry: The Unix Shell](https://swcarpentry.github.io/shell-novice/)
- [Software Carpentry: Programming with Python](https://swcarpentry.github.io/python-novice-inflammation/)
- [Software Carpentry: R for Reproducible Scientfic Analysis](https://swcarpentry.github.io/r-novice-gapminder/) 
- [Software Carpentry: Snakemake for Bioinformatics](https://carpentries-incubator.github.io/snakemake-novice-bioinformatics/)
- [The Caprentries Incubator: Getting Started with Mamba](https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/01-getting-started-with-conda/index.html)  
- [HPC Carpentry](https://www.hpc-carpentry.org/hpc-shell/setup.html)  
- [The Biologists' Guide to Computing](https://www.mrc-lmb.cam.ac.uk/scicomp-new/biologists-guide-to-computing-stable/index.html)

## Schedule  

+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| Week        | Topic                                      | Assignments                                                                                                         |
+=============+============================================+=====================================================================================================================+
| 1/12 - 1/16 | What is Genomics?                          | - [**Preliminary Assessment**](assessment.qmd)                                                                      |
|             |                                            | - [Hohenlohe et al. 2020](https://doi.org/10.1111/mec.15720) [(**pdf**)](papers/hohenlohe_et_al_2020.pdf)           |
|             |                                            | - [Felsenstein rant #1](https://felsenst.github.io/rants.html)                                                      |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 1/19 - 1/23 | Markdown & the Command Line                | - [**Markdown & Command Line Basics (HW1)**](cli.qmd)                                                               |
|             |                                            | - [Sandve et al. 2013](https://doi.org/10.1371/journal.pcbi.1003285) [(**pdf**)](papers/sandve_et_al_2013.pdf)      |
|             |                                            | - [Brandies & Hogg 2021](https://doi.org/10.1371/journal.pcbi.1008645) [(**pdf**)](papers/brandies_hogg_2021.pdf)   |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 1/26 - 1/30 | Version Control with Git & Github          | - [**Git & GitHub Basics (HW2)**](git.qmd)                                                                          |
|             |                                            | - [Blischak et al. 2016](https://doi.org/10.1371/journal.pcbi.1004668) [(**pdf**)](papers/bilschak_et_al_2016.pdf)  |
|             |                                            | - [Braga et al. 2023](https://doi.org/10.1111/2041-210X.14108) [(**pdf**)](papers/braga_et_al_2023.pdf)             |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 2/2 - 2/6   | Computing Clusters                         | - [**Introduction to Tempest (HW3)**](hpc.qmd)                                                                      |
|             |                                            | - [Alnasir 2021](https://doi.org/10.1371/journal.pcbi.1009207) [(**pdf**)](papers/alnasir_2021.pdf)                 |
|             |                                            | - [Tempest Documentation](https://www.montana.edu/uit/rci/tempest/)                                                 |
|             |                                            | - [Globus Documentation](https://www.montana.edu/uit/rci/globus/)                                                   |
|             |                                            | - [MSU VPN Documentation](https://www.montana.edu/uit/computing/desktop/vpn/)                                       |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 2/9 - 2/13  | Sequencing Strategies and                  | - [**Filtering .`fastq` Files (HW4)**](reads.qmd)                                                                   |
|             | Short Read Quality Control                 | - [Illumina Sequencing-by-Synthesis Video](https://www.youtube.com/watch?v=EDVKxSNdSic)                             |      
|             |                                            | - [Fuentes-Pardo & Ruzzante 2017](https://doi.org/10.1111%2Fmec.14264) [(**pdf**)](papers/fuentes-pardo_2017.pdf)   |
|             |                                            | - [Toews et al. 2016 (optional)](https://doi.org/10.1642/AUK-15-51.1) [(**pdf**)](papers/toews_et_al_2016.pdf)      |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 2/16 - 2/20 | Reference Genomes & Alignment              |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 2/23 - 2/27 | Variant Calling                            |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 3/2 - 3/6   | Variant Filtering                          |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 3/9 - 3/13  | Workflow Managers                          |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 3/16 - 3/20 | Spring Break                               |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 3/23 - 3/27 | Kinship and Inbreeding                     |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 3/30 - 4/3  | Genetic Diversity and Population Structure |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 4/6 - 4/10  | Introgression and Hybridization            |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 4/13 - 4/17 | Phylogenetics and Conservation Units       |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 4/20 - 4/24 | Simulations and Demographic History        |                                                                                                                     |
|             |                                            |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 4/27 - 5/1  | Detecting Natural Selection                |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+
| 5/4 - 5/8   | Finals Week                                |                                                                                                                     |
+-------------+--------------------------------------------+---------------------------------------------------------------------------------------------------------------------+

## Drop / Add Policy

MSU’s registration processes can be found on the Registrar’s website. January 27th is the last day to drop online; February 3rd is the last day to drop without a ‘W’ on your transcript; April 15th is the last day to drop with a ‘W’ on your transcript.

## Generative AI

Generative AI tools are permitted to help debug code; they are *not* permitted as a shortcut to learning functions or commands from software documentation, and *especially not* permitted for summarizing reading assignments, where the point of the exercise is to become comfortable interpreting the technical genomics literature. I will not spend my time actively policing their use, but if detected, I will consider it academic misconduct.

## Attendance Policy

Please do not come to campus if you are sick! I trust you will only miss class when absolutely necessary, and do not need extensive explanations for absences. However, please notify me as soon as practical, so that we can figure appropriate accommodations.

## Inclusivity Statement

I support an inclusive learning environment where diversity and individual differences are understood, respected, appreciated, and recognized as a source of strength. We expect that students, faculty, administrators and staff at MSU will respect differences and demonstrate diligence in understanding how other peoples' perspectives, behaviors, and worldviews may be different from their own.

## Syllabus Language for Students with Disabilities

If you are a student with a disability and wish to use your approved accommodations for this course, please contact me during my office hours to discuss. Please have your Accommodation Notification or Blue Card available for verification of accommodations. Accommodations are approved through the Office of Disability Services located in SUB 174. Please see Disability Services for more information.  
