---
format: html
---

# Understanding `.fastq` Data and Short Read Quality Control

Our [introduction to the Tempest HPC](hpc.qmd) included our first exposure to real live short read sequencing data. I will assume that this week's reading and discussion has everyone feeling a little more comfortable with how these data are generated and what a read is in the first place. Today's lab, which is somewhat shorter than its predecessors, will cover the basics of `.fastq` data: its format, what aspects of a sample and its preparation are contained in reads, and how to trim and filter reads to ensure that only high-quality data get passed on to the next stage of your workflow. 

## Example Data

The remaining classes before spring break will lead us through an abbreviated sequence alignment, variant calling, and variant filtering pipeline. These basic bioinformatic processing steps are required for nearly all population-level genomics and transcriptomics projects; while the exact programs and filtering choices will vary, the backbone presented here should orient you to successfully preparing your own data for subsequent analysis. To achieve this goal, we will work with unpublished exon capture data from 13 species of high Andean tanagers (Passeriformes: Thraupidae). These data were generated by exon capture---a kind of reduced representation library preparation method that targets specific expressed genes by enriches them with synthetic user-designed probes---of the seven hemoglobin peptide chains, or subunits. Because hemoglobin structure impacts its ability to bind and transport $O_2$ in the blood, variation in underlying sequences may be associated with adaptations to high or low altitude and attendant difference in the partial pressure of oxygen ($PO_2$). 

Below is a complete list of species with linked natural history and distributional data. Feel free to pick your favorite, but note that sample sizes vary, posing a trade-off between the time needed to run tools and how interesting the data ultimately is. All can be found in named subdirectories within `bioe-591-genomics/course-materials/data/raw_reads/`. 

- [Giant Conebill *Oreomanes fraseri*](https://birdsoftheworld.org/bow/species/giacon1/cur/introduction) (outdated binomial used here)
- [Capped Conebill *Conirostrum albifrons*](https://birdsoftheworld.org/bow/species/capcon1/cur/introduction)
- [Cinereous Conebill *Conirostrum cinereum*](https://birdsoftheworld.org/bow/species/cincon1/cur/introduction)
- [Blue-backed Conebill *Conirostrum sitticolor*](https://birdsoftheworld.org/bow/species/blbcon1/cur/introduction)
- [White-sided Flowerpiercer *Diglossa albilatera*](https://birdsoftheworld.org/bow/species/whsflo1/cur/introduction)
- [Cinnamon-bellied Flowerpiercer *Diglossa baritula*](https://birdsoftheworld.org/bow/species/cibflo1/cur/introduction)
- [Black-throated Flowerpiercer *Diglossa brunneiventris*](https://birdsoftheworld.org/bow/species/bktflo1/cur/introduction)
- [Blueish Flowerpiercer *Diglossa caerulescens*](https://birdsoftheworld.org/bow/species/bluflo1/cur/introduction)
- [Masked Flowerpiercer *Diglossa cyanea*](https://birdsoftheworld.org/bow/species/masflo1/cur/introduction)
- [Deep-blue Flowerpiercer *Diglossa glauca*](https://birdsoftheworld.org/bow/species/debflo1/cur/introduction)
- [Glossy Flowerpiercer *Diglossa lafresnayii*](https://birdsoftheworld.org/bow/species/gloflo1/cur/introduction)
- [Moustached Flowerpiercer *Diglossa mystacalis*](https://birdsoftheworld.org/bow/species/mouflo1/cur/introduction)
- [Rusty Flowerpiercer *Diglossa sittoides*](https://birdsoftheworld.org/bow/species/rusflo1/cur/introduction)

##  The `.fastq` Format 

Sequencing reads are turned into plain-text data as `.fastq` files, which are then typically compressed using `gzip` into compressed `.fast.gz` files. Typing `gunzip filename.fastq.gz` will decompress the target, albeit slowly; because most tools can read compressed data and even small genomic datasets can be prohibitively large in their raw form, there is usually little reason to do this. However, it can be useful to look at least part of a human-readable `.fastq`. Luckily, we can do so using a version of the familiar `bash` command `cat` that works on compressed data (`zcat`). We'll pipe this to `head` to avoid printing the entire contents of the file to your screen; if you'd like to expand beyond ten lines, use `head`'s `-n <integer>` argument.  

```bash
zcat intro/catamenia_analis.fastq.gz | head
```

The output of this command should look something like this: 

```bash
@J00138:88:HG7TGBBXX:3:1101:1083:1086 1:N:0:NGCGGAAT+NATGTTCC
NTCAAAGTCCAGCCCAGACCTCTCACCTTGCAGACAACTCCTTCCCCAACACGGCCCCCCACACCCCGCCCCCCCCAACCCCCCCCCAAACCCCCCAACAGCCCCCCCCGGCCCCCCCCTGCTCCCGCCCCCCCCCCCCGGCTCCCCCCCC
+
#<AAFJFAFFJ77JJ<JJJJJJJ<JJJJJJJ7J7AJ7JJJJAF<FAA--F-7--7-AAJ----777-7A-7FJA----777JJ----7--7-AJF-------7-7AA----7-FF7A77-----------A-A)-A)-)))))))-A---)
@J00138:88:HG7TGBBXX:3:1101:7415:1086 1:N:0:NGCGGAAT+NATGTTCC
NTCAGGGCATGGGCAAATGCATAGACTGGCACAGACCCTCCCCAGAACTGGGAATTCAGGGCACGGGCAAATGCACAGACTGGCACAGACCCTCCCCAGAACTGGGAATTCAGGGCATGGGCAAATGCACAGACTGGCACAGACCCTCCCC
+
#AAAFFAJJJJJJJFJJJJJJJJJJJJJFFJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJFAJJJAJJJJJJFJJFJFJJJJJJJFJA<JAJJJJJJJJFJJJJJFJFFJFJJFJF-AFFJAFJFFJJJA-<AFJ-FAFFJJA7-)<A<AJ
@J00138:88:HG7TGBBXX:3:1101:8958:1086 1:N:0:NGCGGAAT+NATGTTCC
NGTGAGCGAGCAACAGCGCCAGAGGGCAGACAGGATGGAAGAGCGCAAGAACAGAGCCTGGGGAGTGCGGAGACGCCGGCAGGCAGTTCACTTCCCATGCTCTTTCTTCTCCTGTGTGTCGTGACAGCAGCAGCAGCAGCAGCAGCAGCAG
```

Though initially overwhelming, you will notice that the characters above consist of only four lines, which can be interpreted as follows: 

1) a unique sequence identifier; 
2) the nucleotide sequence itself; 
3) a quality score identifier line (simply a `+`)
4) a quality score, encoded by an `ASCII` character. 

The sequence identifier, which is initially cryptic, follows the format below. Note that these are technical details about sequencing itself, and largely not the biology of the sample. 

```bash
@<instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<x-pos>:<y-pos> <read>:<is filtered>:<control number>:<sample number>
```

The nucleotide sequence itself should be self explanatory, though note that `N` (and sometimes other )

(Illumina [has a helpful explainer](https://help.basespace.illumina.com/files-used-by-basespace/fastq-files) on exactly what each character preceding the sequence data means.) 

In your own subdirectory, dig up the `size.txt` file you generated during last week's activity. 

```bash 
reads	bases
5417329	818016679
```